### PR TITLE
add support for background notifications

### LINKED ISSUE #317: add support for background notifications(implement firestore cloud messaging)

### PR DESCRIPTION
**Pull request by BogdanCDev**
## Summary

This PR enhances the notification system to support background notifications, ensuring users receive friend request alerts and other messages even when the app is in background. This is a new feature to improve the user experience by leveraging background services effectively.

## Changes

- **Notification Handling:**
  - Updated `VoyageurMessagingService` to process Firebase Cloud Messaging (FCM) data messages for background notifications.
  - Enhanced `AndroidNotificationProvider` to manage and display friend request notifications triggered from background states.
  - Refined `NotificationHelper` to support notifications with contextual details based on app state.
- **Background Support:**
  - Added lifecycle-aware token updates in `VoyageurMessagingService` for improved user session handling.
  - Ensured notifications are displayed even when the app is closed or minimized.
  - Implemented logic to open the app in the correct state when a notification is pressed, such as navigating to specific screens or content based on the notification type.

## Checklist

- [X] This PR resolves #317
- [X] Tests have been added for new functionality.
- [X] Screenshots or UI changes have been attached 
- [X] Documentation has been updated 

##  Testing

Explain how this feature was tested and what kind of testing (unit, integration, UI) has been done. Mention platforms or configurations used for testing.
- **Manual Testing:**
    - [X] Tested on emulator / physical device.
- **Unit Tests**
    - [X] Added units tests for this

##  Related Issues/Tickets

Closes #317

## Screenshots

https://github.com/user-attachments/assets/dd3c7ee0-2af1-4488-9794-431426e43434

## 💡 Additional Context

Testing for these changes is extremely complicated.

### CODE REVIEW

**Review by apana03 — COMMENTED** [review]
Thank you for working on this complex feature Bogdan! I had a few remarks that I think should be addressed before finalizing this PR. Regarding the testing, I understand it can be difficult to achieve the required line coverage but I still think we should aim for it. However, if it becomes to time consuming an exception can be made.

File: app/src/main/java/com/android/voyageur/MainActivity.kt [hunk 1]
```diff
@@ -85,17 +75,80 @@ class MainActivity : ComponentActivity() {
     }
   }
 
+  private fun initializeFirebase() {
+    try {
+      FirebaseApp.initializeApp(this)
+      getFCMToken()
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/MainActivity.kt:81]: Great suggestion!

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 2]
```diff
@@ -315,6 +297,62 @@ constructor(
...
+  }
+
+  override fun onCleared() {
+    super.onCleared()
+    if (addAuthStateListener) {
+      firebaseAuth.removeAuthStateListener(authStateListener)
+    }
+    userListenerRegistration?.remove()
+    sentFriendRequestsListener?.remove()
+    friendRequestsListener?.remove()
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:312]: I don't believe it is necessary in this instance, but you are right, better to be on the safer side.

File: app/src/main/java/com/android/voyageur/MainActivity.kt [hunk 3]
```diff
@@ -85,17 +75,80 @@ class MainActivity : ComponentActivity() {
     }
   }
 
+  private fun initializeFirebase() {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/MainActivity.kt:78]: Done

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 4]
```diff
@@ -315,6 +297,62 @@ constructor(
     }
   }
 
+  // Add this function to manually trigger token update
+  fun refreshFCMToken() {
+    firebaseAuth.currentUser?.uid?.let { userId -> updateStoredFCMToken(userId) }
+  }
+
+  override fun onCleared() {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:305]: done

File: app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt [hunk 5]
```diff
@@ -624,9 +669,28 @@ constructor(
           Log.e("FRIEND_REQUEST", "Failed to delete friend request: ${exception.message}")
         })
   }
+
+  fun updateFriendRequest(friendRequest: FriendRequest, accepted: Boolean) {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/model/user/UserViewModel.kt:673]: done

File: app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+  override fun onMessageReceived(remoteMessage: RemoteMessage) {
+    super.onMessageReceived(remoteMessage)
+    Log.d(TAG, "From: ${remoteMessage.from}")
+
+    val notificationProvider = AndroidNotificationProvider(this)
+
+    remoteMessage.data.let { data ->
+      when (data["type"]) {
+        "friend_request" -> {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt:35]: done

File: app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  private val TAG = "VoyageurMessagingService"
+  private val db = FirebaseFirestore.getInstance()
+
+  override fun onCreate() {
+    super.onCreate()
+    val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+    NotificationHelper.createNotificationChannel(this, notificationManager)
+  }
+
+  override fun onMessageReceived(remoteMessage: RemoteMessage) {
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt:27]: done

File: app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        "friend_request_accepted" -> {
+          val acceptorName = data["acceptorName"] ?: getString(R.string.unknown)
+          notificationProvider.showFriendRequestAcceptedNotification(acceptorName)
+        }
+      }
+    }
+
+    remoteMessage.notification?.let { notification ->
+      Log.d(TAG, "Message Notification Body: ${notification.body}")
```
[the remark this answers is not in the document]
Comment (BogdanCDev) [reply app/src/main/java/com/android/voyageur/services/VoyageurMessagingService.kt:77]: Thank you very much!

**Review by apana03 — APPROVED** [review]
Nice work Bogdan! Thank you for taking into account my comments. Once the CI passes feel free to merge your changes :)

**Pull request merged**